src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs

```diff
@@ -45,19 +44,25 @@ private abstract class CompilationTrackerState
                 /// empty, otherwise a <see cref="InProgressState"/>.
                 /// </summary>
                 public static WithCompilationTrackerState Create(
+                    bool isFrozen,
                     Compilation compilationWithoutGeneratedDocuments,
                     CompilationTrackerGeneratorInfo generatorInfo,
                     Compilation? compilationWithGeneratedDocuments,
                     ImmutableList<(ProjectState state, CompilationAndGeneratorDriverTranslationAction action)> intermediateProjects)
                 {
                     Contract.ThrowIfTrue(intermediateProjects is null);
 
+                    // If we're not frozen, transition back to the non-final state as we def want to rerun generators
+                    // for either of these non-final states.
+                    if (!isFrozen)
+                        generatorInfo = generatorInfo.WithDocumentsAreFinal(false);
```

Comment:
note that this was the special logic within generatorInfo to not move to the non-final state if we were frozen.  also note that this is the only location in the code where we set WithDocumentsAreFinal(False).  so it was clearer to just have teh logic be here to me versus having a sepcial method with special logic that was only called in a single location.